std/typecons.d

```diff
@@ -2094,6 +2100,343 @@ enum isTuple(T) = __traits(compiles,
+        static assert(rv.length == Types.length,
+            "critical: length mismatch in Tie; this is a bug");
+        alias rvalues = rv;
+        Types values;
+        alias values this;
+    pure nothrow @nogc @safe @disable:
+        void opAssign(T...)(T);
+        this(this);
+    }
+
+    enum bool isRvalue(alias value) = !__traits(isRef, value);
+}
+
+unittest
+{
+    const int x;
+    int y;
+    tie[ y ] = tie[ x ];
+    tie[ y ] = tuple(x);
+    tie[ y ] = cast(const(int)[1]) [ x ];
+    tie[ y ] = cast(const(int[1])) [ x ];
+    const(int)[1] icar = [ x ];
+    tie[ y ] = icar;
+    const(int[1]) iarc = [ x ];
+    tie[ y ] = iarc;
+}
+
+/// Typical use cases.
+pure nothrow @nogc @safe
+unittest
+{
+    int x = 2, y = 3;
+
+    // Swapping.
+    tie[ x, y ] = tie[ y, x ];
+    assert(x == 3 && y == 2);
+    // Simultaneous iteration.
+    tie[ x, y ] = tie[ y + 1, x * 2 ];
+    assert(x == 3 && y == 6);
+    // If all types support op=, you can use op= on `tie`.
+    tie[ x, y ] += tie[ 2, 1 ];
+    assert(x == 5 && y == 7);
+
+    // $ is used for discarding.
+    tie[ $, x, $, y, $ ] = tie[ 0, 1, 2, 3, 4 ];
+    assert(x == 1 && y == 3);
+
+    static auto returnsTuple(int v) { return tuple(v * 2, v + 2); }
+    static int[3] returnsStaticArray(int v) { return [ v - 1, v, v + 1 ]; }
+
+    // Deconstruction.
+    tie[ x, y ] = returnsTuple(3);
+    assert(x == 6 && y == 5);
+    tie[ x, $, y ] = returnsStaticArray(x);
+    assert(x == 5 && y == 7);
+
+    // Errors with messages:
+    // Length mismatch.
+    static assert(!__traits(compiles, {
+        tie[ x, y ] = tie[ 1, 2, 3];
+    }));
+    // Left-hand side objects must be lvalues or the constant null or $.
+    static assert(!__traits(compiles, {
+        tie[ 1, x ] = tie[ y, 2 ];
+    }));
+    // Assigned value must be a tie, a std.typecons.Tuple or a static array; this is a int[]
+    static assert(!__traits(compiles, {
+        tie[ x, y ] = [ 1, 2 ]; // [ 1, 2 ] is a dynamic array.
+    }));
+    tie[ x, y ] = cast(int[2]) [ 1, 2 ]; // cast necessary
+    assert(x == 1 && y == 2);
+}
+
+/// Even complicated usages can be executed at compile-time.
+pure nothrow @safe
+unittest
+{
+    static string ctfeFunction() pure nothrow @safe
+    {
+        int[] ar = [ 1, 2, 3, 4 ];
+        string s = "Hello";
+        // Simultaneous iteration.
+        tie[ ar, s ] ~= tie[ cast(int) s.length, ", World!?"[0 .. ar[$ - 1] * 2] ];
+        assert(ar == [ 1, 2, 3, 4, 5 ] && s == "Hello, World!");
+
+        // op= with discard.
+        auto arx = [ 6, 7 ];
+        tie[ ar, $, s ] ~= tie[ arx, true, '?' ]; // `true` being discarded
+        import std.range : drop;
+        assert(ar.drop(5) == [ 6, 7 ]);
+
+        return s[0 .. $ - 1];
+    }
+    enum string hw = ctfeFunction(); // evaluated at compile-time
+    static assert(hw == "Hello, World!");
+}
+
+/// `tie` supports moving. It can be observed using an explicit destructor.
+pure nothrow @nogc @safe
+unittest
+{
```

Comment:
Thats one long unittest. Can it be split?

Reply:
I have split it.